File: src/content/3.5/code/scala/snippet29.scala
implicit def contMonad[R] = new Monad[Cont[R, ?]] {
  def flatMap[A, B](ka: Cont[R, A])(kab: A => Cont[R, B]): Cont[R, B] =
    Cont { hb =>
      runCont(ka)(a => runCont(kab(a))(hb))
    }

  def pure[A](a: A): Cont[R, A] =
    Cont(ha => ha(a))
}

File: src/content/3.5/code/ocaml/snippet29.ml
module Cont_Monad (R : sig
  type t
end) : Monad_Bind = struct
  type 'a m = (R.t, 'a) cont

  let return a = Cont (fun ha -> ha a)

  let ( >>= ) ka kab =
    Cont (fun hb -> run_cont ka (fun a -> run_cont (kab a) hb))
end
